### ISSUE TITLE
Unable to recover etcd after loss of quorum

### ISSUE THREAD

**Issue by tmitchell** [rancher/rke2#2150]
**Environmental Info:**
RKE2 Version:  
v1.20.12+rke2r1

Node(s) CPU architecture, OS, and Version: 
x86_64 RHEL8

Cluster Configuration:
3 servers, 5 agents

**Describe the bug:**
We had a loss of quorum while doing an upgrade of the servers.  Because we were upgrading using new nodes, we have no snapshot to restore from.  Ran a `systemctl stop rke2-server; rke2 server --cluster-reset` as described in the docs on one of the servers.  

**Steps To Reproduce:**
- Installed RKE2 v1.20.8
- Upgraded servers to 1.20.12
- Loss of quorum due to ASG mishap
- Followed cluster reset instructions from https://docs.rke2.io/backup_restore/

**Expected behavior:**
etcd restored 

**Actual behavior:**
The `rke2 server --cluster-reset` command never completes, it just gets stuck in a loop with:
```
INFO[0190] Waiting for etcd server to become available
INFO[0193] Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:9345/v1-rke2/readyz: 500 Internal Server Error
INFO[0193] Waiting for kubelet to be ready on node ip-10-XX-XX-XX.us-west-1.compute.internal: Get "https://127.0.0.1:6443/api/v1/nodes/ip-10-XX-XX-XX.us-west-1.compute.internal": dial tcp 127.0.0.1:6443: connect: connection refused
WARN[0198] Unable to watch for tunnel endpoints: Get "https://127.0.0.1:6443/api/v1/namespaces/default/endpoints?fieldSelector=metadata.name%3Dkubernetes&resourceVersion=0&watch=true": dial tcp 127.0.0.1:6443: connect: connection refused
INFO[0198] Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:9345/v1-rke2/readyz: 500 Internal Server Error
{"level":"warn","ts":"2021-11-19T00:13:59.795Z","caller":"clientv3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"passthrough:///https://127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = latest balancer error: connection error: desc = \"transport: Error while dialing dial tcp 127.0.0.1:2379: connect: connection refused\""}
INFO[0200] Failed to test data store connection: context deadline exceeded
WARN[0203] Unable to watch for tunnel endpoints: Get "https://127.0.0.1:6443/api/v1/namespaces/default/endpoints?fieldSelector=metadata.name%3Dkubernetes&resourceVersion=0&watch=true": dial tcp 127.0.0.1:6443: connect: connection refused
INFO[0203] Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:9345/v1-rke2/readyz: 500 Internal Server Error
WARN[0208] Unable to watch for tunnel endpoints: Get "https://127.0.0.1:6443/api/v1/namespaces/default/endpoints?fieldSelector=metadata.name%3Dkubernetes&resourceVersion=0&watch=true": dial tcp 127.0.0.1:6443: connect: connection refused
INFO[0208] Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:9345/v1-rke2/readyz: 500 Internal Server Error
WARN[0213] Unable to watch for tunnel endpoints: Get "https://127.0.0.1:6443/api/v1/namespaces/default/endpoints?fieldSelector=metadata.name%3Dkubernetes&resourceVersion=0&watch=true": dial tcp 127.0.0.1:6443: connect: connection refused
INFO[0213] Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:9345/v1-rke2/readyz: 500 Internal Server Error
{"level":"warn","ts":"2021-11-19T00:14:14.796Z","caller":"clientv3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"passthrough:///https://127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = latest balancer error: connection error: desc = \"transport: Error while dialing dial tcp 127.0.0.1:2379: connect: connection refused\""}
INFO[0215] Failed to test data store connection: context deadline exceeded
WARN[0218] Unable to watch for tunnel endpoints: Get "https://127.0.0.1:6443/api/v1/namespaces/default/endpoints?fieldSelector=metadata.name%3Dkubernetes&resourceVersion=0&watch=true": dial tcp 127.0.0.1:6443: connect: connection refused
INFO[0218] Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:9345/v1-rke2/readyz: 500 Internal Server Error
```

**Additional context / logs:**
Digging into the failing etcd container via crictl yields the following:
```
# ./bin/crictl  --runtime-endpoint unix:///run/k3s/containerd/containerd.sock logs f8492d5bec2da
2021-11-19 00:13:47.395916 I | etcdmain: Loading server configuration from "/var/lib/rancher/rke2/server/db/etcd/config". Other configuration command line flags and environment variables will be ignored if provided.
2021-11-19 00:13:47.395956 E | etcdmain: error verifying flags, open /var/lib/rancher/rke2/server/db/etcd/config: permission denied. See 'etcd --help'.
```
A comparison of the permissions with another working cluster shows no deviation.

**Comment by brandond** [comment]
`2021-11-19 00:13:47.395956 E | etcdmain: error verifying flags, open /var/lib/rancher/rke2/server/db/etcd/config: permission denied. See 'etcd --help'.`

I believe in the past we've seen this when selinux policy denies the pods access to the files. Did your new nodes by any chance have updated selinux/container-selinux packages installed on them?

**Comment by tmitchell** [comment]
Indeed.

Before: `container-selinux.noarch                      2:2.164.1-1.module+el8.4.0+11870+8b6f7018   @rhel-8-appstream-rhui-rpms`
After: `container-selinux.noarch                      2:2.167.0-1.module+el8.5.0+12582+56d94c81 @rhel-8-appstream-rhui-rpms`

Why would this crop up only after doing the reset?  When the new nodes joined etcd was happy enough.

**Comment by brandond** [comment]
You might try installing the latest rke2-selinux package from https://github.com/rancher/rke2-selinux/releases/tag/v0.9.testing.1

**Comment by brandond** [comment]
You should install that package as an upgrade, not a new install. Check your rpm flags. I suspect once you get that resolved you'll be able to enable selinux again.

**Comment by tmitchell** [comment]
Roger.  Not enough coffee, got the package installed.  It seems I spoke too soon however.  

The cluster-reset finally went through, and I proceeded to start things back up, however etcd was still in a non-functional state (unfortunately didn't grab logs).  I tried to run the restore once more and I'm back to the original error with 
`etcdmain: error verifying flags, open /var/lib/rancher/rke2/server/db/etcd/config: permission denied. See 'etcd --help'.`

Further, checking the audit logs doesn't show any selinux denies so now I'm really at a loss as to what's going on (and why the reset seemed to run but not actually restore things).

**Comment by brandond** [comment]
Is your cluster hardened (running with --profile=X)? Did you re-enable selinux after updating the rke2-selinux package?

**Comment by tmitchell** [comment]
Yes running with CIS 1.5.   Tried with selinux both enforcing and permissive.

**Comment by brandond** [comment]
Hmm. You may need to ensure that both rke2-selinux and container-selinux are up to date, and then run a full restorecon on /var/lib/rancher.

Any other ideas @dweomer ?

**Comment by mitchellmaler** [comment]
I ran into the same issue. To complete the reset, I just commented out the profile and selinux: true in the config and switched se to permissive. Once the reset was successful (says to start services, etc.) I just enabled the profile and selinux support along with switching back to enforcing. That seemed to do the trick. Didn't have to run a restorecon on /var/lib/rancher directory and the regular services/containers started up fine.
